src/protocols/fiat-protocol.js

```diff
@@ -0,0 +1,217 @@
+// limitations under the License.
+'use strict'
+
+import { NotImplementedError } from '../errors.js'
+
+/** @typedef {import('../wallet-account-read-only.js').IWalletAccountReadOnly} IWalletAccountReadOnly */
+
+/** @typedef {import('../wallet-account.js').IWalletAccount} IWalletAccount */
+
+/**
+ * Standardized status for an on/off-ramp transaction.
+ * @typedef {'in_progress' | 'failed' | 'completed'} FiatTransactionStatus
+ */
+
+/**
+ * A protocol-agnostic, standardized object representing the details of an on/off-ramp transaction.
+ * @typedef {Object} FiatTransactionDetail
+ * @property {FiatTransactionStatus} status - The current status of the transaction.
+ * @property {string} cryptoAsset - The provider-specific code of the crypto asset (e.g., 'btc').
+ * @property {string} fiatCurrency - The currency's ISO 4217 code (e.g., 'USD').
+ * @property {Record<string, any>} [metadata] - Provider-specific raw data for this transaction.
+ */
+
+/**
+ * A protocol-agnostic, standardized object representing a supported crypto asset.
+ * @typedef {Object} FiatSupportedAsset
+ * @property {string} code -Provider-specific asset code for the crypto asset.
+ * @property {string} networkCode - The network code for the asset, if applicable (e.g., 'ethereum', 'tron').
+ * @property {number} precision - The number of decimal places for the asset.
+ * @property {string} [name] - The asset's full name (e.g., 'Bitcoin').
+ * @property {Record<string, any>} [metadata] - Provider-specific raw data for this asset.
+ */
+
+/**
+ * A protocol-agnostic, standardized object representing a supported fiat currency.
+ * @typedef {Object} FiatSupportedCurrency
+ * @property {string} code - The currency's ISO 4217 code (e.g., 'USD').
+ * @property {number} precision - The number of decimal places for the currency.
+ * @property {string} [name] - The currency's full name (e.g., 'United States Dollar').
+ * @property {Record<string, any>} [metadata] - Provider-specific raw data for this currency.
+ */
+
+/**
+ * A protocol-agnostic, standardized object representing a supported country.
+ * @typedef {Object} FiatSupportedCountry
+ * @property {string} code - The country's ISO 3166-1 alpha-2 or alpha-3 code.
+ * @property {boolean} isBuyAllowed - Whether buying is supported in this country.
+ * @property {boolean} isSellAllowed - Whether selling is supported in this country.
+ * @property {string} [name] - The country's common name.
+ * @property {Record<string, any>} [metadata] - Provider-specific raw data for this region.
+ */
+
+/**
+ * @interface
+ */
+export class IFiatProtocol {
+  /**
+   * Generates a widget URL for a user to purchase a crypto asset with fiat currency.
+   * @param {string} cryptoAsset - The provider-specific code of the crypto asset to purchase.
+   * @param {string} fiatCurrency - The currency's ISO 4217 code (e.g., 'USD').
+   * @param {number} amount - The amount of crypto asset to buy, in its main unit (e.g., 1.50 for 1.50 ETH).
```

Comment:
This amount should definetely be in base unit rather than main unit in order to maintain consistency across all wallet and protocol methods and avoid precision errors. The type should also accept big integers (as all other token amount arguments do).

```suggestion
   * @param {number | bigint} amount - The amount of crypto asset to buy, in its base unit (e.g., 1,500,000,000,000,000,000 for 1.50 ETH).
```

Reply:
i thought about this but unfortunately the amount passed to the widget URL is in main unit. if we pass base unit, then we would need to handle the conversion as well (likely need to pass precision in). that's why i changed `amount` to `number`, i was going for `number | bigint` at first

Reply:
digging deeper into this, MoonPay only understands and returns data according to main unit. here's a metadata of USDC  from MoonPay, USDC has precision of 6 but here it says 2, meaning MoonPay accepts numbers like 0.05

```
{
    "id": "aaefa32f-161b-42c8-8115-debcbf3d6a2d",
    "createdAt": "2019-05-17T18:24:45.206Z",
    "updatedAt": "2019-05-17T18:24:45.206Z",
    "type": "crypto",
    "name": "USD Coin (ERC-20)",
    "code": "usdc",
    "precision": 2,
    "minBuyAmount": 30,
    "maxBuyAmount": 1000,
    "minSellAmount": 15,
    "maxSellAmount": 100000,
    ...
}
```